src/mscorlib/shared/System/Text/StringBuilder.cs

```diff
@@ -2001,22 +2043,23 @@ private static unsafe void ThreadSafeCopy(char* sourcePtr, char[] destination, i
             }
         }
 
-        private static void ThreadSafeCopy(char[] source, int sourceIndex, char[] destination, int destinationIndex, int count)
+        private static unsafe void ThreadSafeCopy(char[] source, int sourceIndex, Span<char> destination, int destinationIndex, int count)
         {
             if (count > 0)
             {
-                if ((uint)sourceIndex <= (uint)source.Length && (sourceIndex + count) <= source.Length)
+                if ((uint)sourceIndex > (uint)source.Length || (sourceIndex + count) > source.Length)
                 {
-                    unsafe
-                    {
-                        fixed (char* sourcePtr = &source[sourceIndex])
-                            ThreadSafeCopy(sourcePtr, destination, destinationIndex, count);
-                    }
+                    throw new ArgumentOutOfRangeException(nameof(sourceIndex), SR.ArgumentOutOfRange_Index);
                 }
-                else
+
+                if ((uint)destinationIndex > (uint)destination.Length || (destinationIndex + count) > destination.Length)
```

Comment:
FWIW, the same condition exists on the other ThreadSafeCopy overload (that I didn't touch), line 2034.
Still, I agree on the risk. Gonna do some further testing.